### PR TITLE
[Merged by Bors] - refactor(AlgebraicGeometry): Introduce `HasRingHomProperty`.

### PR DESCRIPTION
**Pull request by erdOne**
`HasRingHomProperty P Q` is a type class asserting that `P` is local at the target and the source,
and for `f : Spec B ⟶ Spec A`, it is equivalent to the ring hom property `Q` on `Γ(f)`.
This replace the previous API `affineLocally` that is both slow and hard to use.

This is then used to refactor `LocallyOfFiniteType` and to deduplicate code.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by erdOne** [comment]
!bench

File: Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.lean [hunk 1]
```diff
@@ -27,8 +27,31 @@ the target. (TODO)
...
 For the latter `P` should be local on the target (`RingHom.PropertyIsLocal P`), and
 `affineLocally P` will be local on both the source and the target.
-
-Further more, these properties are stable under compositions (resp. base change) if `P` is. (TODO)
+We also provide the following interface:
+
+## `HasRingHomProperty`
+
+`HasRingHomProperty P Q` is a type class asserting that `P` is local at the target and the source,
+and for `f : Spec B ⟶ Spec A`, it is equivalent to the ring hom property `Q`.
```
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.lean:35]:
```suggestion
and for `f : Spec B ⟶ Spec A`, it is equivalent to the ring hom property `Q` evaluated on `Γ(f)`.
```

File: Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.lean [hunk 2]
```diff
@@ -38,504 +61,311 @@ universe u
...
+    fun i ↦ iff_of_openCover (P := P) (f := 𝒰.map i ≫ f) (𝒰.obj i).affineCover]
+  simp [Scheme.OpenCover.affineRefinement]
+
+instance : P.ContainsIdentities where
+  id_mem X := by
+    rw [IsLocalAtTarget.iff_of_iSup_eq_top (P := P) _ (iSup_affineOpens_eq_top _)]
+    intro U
+    rw [morphismRestrict_id, iff_of_isAffine (P := P), Scheme.id_app]
+    have := IsLocalization.at_units (.powers (1 : Γ(X ∣_ᵤ 𝟙 X ⁻¹ᵁ U, ⊤))) (by simp)
+    exact (isLocal_ringHomProperty P).HoldsForLocalizationAway Γ(X ∣_ᵤ 𝟙 X ⁻¹ᵁ U, ⊤) 1
```
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.lean:300]: Maybe factor this out as a lemma?
```lean
lemma RingHom.id_of_holdsForLocalizationAway (h : RingHom.HoldsForLocalizationAway Q)
    (A : Type u) [CommRing A] : Q (RingHom.id A : A →+* A) := by
  have := IsLocalization.at_units (.powers (1 : A)) (by simp)
  exact h A 1
```

File: Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.lean [hunk 3]
```diff
@@ -38,504 +61,311 @@ universe u
...
+  exact H _ ⟨_, V.2.image_of_isOpenImmersion _⟩ _
+
+variable {P f}
+
+lemma iff_appLE : P f ↔ ∀ (U : Y.affineOpens) (V : X.affineOpens) (e), Q (f.appLE U V e) := by
+  rw [eq_affineLocally P, affineLocally_iff_affineOpens_le]
+
+theorem of_openCover [IsAffine Y]
+    (𝒰 : X.OpenCover) [∀ i, IsAffine (𝒰.obj i)] (H : ∀ i, Q ((𝒰.map i ≫ f).app ⊤)) :
+    P f := by
```
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.lean:221]: Maybe the name should indicate that the cover is of the source, e.g. `of_openCover_source`? Maybe we can make the convention that `openCover` alway relates to the target and `openCover_source` to the source.

File: Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.lean [hunk 4]
```diff
@@ -38,504 +61,311 @@ universe u
...
+      rw [← Scheme.OpenCover.pullbackHom_map_assoc]
+      exact this _ _ (IsLocalAtTarget.of_isPullback (.of_hasPullback _ _) hf)
+        (comp_of_isOpenImmersion _ _ _ hg) inferInstance
+    wlog hX : IsAffine X generalizing X
+    · rw [IsLocalAtSource.iff_of_openCover (P := P) X.affineCover]
+      intro i
+      rw [← Category.assoc]
+      exact this _ (comp_of_isOpenImmersion _ _ _ hf) inferInstance
+    rw [iff_of_isAffine (P := P)] at hf hg ⊢
+    exact (isLocal_ringHomProperty P).StableUnderComposition _ _ hg hf
```
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.lean:321]: Nice!

File: Mathlib/AlgebraicGeometry/Scheme.lean [hunk 5]
```diff
@@ -142,6 +142,19 @@ lemma appLE_congr (e : V ≤ f ⁻¹ᵁ U) (e₁ : U = U') (e₂ : V = V')
...
+    (h_app : ∀ U, f.app U ≫ X.presheaf.map
+      (eqToHom congr((Opens.map $h_base.symm).obj U)).op = g.app U) : f = g :=
+  LocallyRingedSpace.Hom.ext _ _ <| SheafedSpace.ext _ _ h_base
+    (TopCat.Presheaf.ext fun U ↦ by simpa using h_app U)
+
+lemma preimage_iSup {ι} (U : ι → Opens Y) : f ⁻¹ᵁ iSup U = ⨆ i, f ⁻¹ᵁ U i :=
+  Opens.ext (by simp)
+
+lemma preimage_iSup_eq_top {ι} {U : ι → Opens Y} (hU : iSup U = ⊤) :
+    ⨆ i, f ⁻¹ᵁ U i = ⊤ := (f.preimage_iSup U).symm.trans (by rw [hU]; rfl)
```
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Scheme.lean:156]:
```suggestion
    ⨆ i, f ⁻¹ᵁ U i = ⊤ := f.preimage_iSup U ▸ hU ▸ rfl
```

File: Mathlib/AlgebraicGeometry/Morphisms/FiniteType.lean [hunk 6]
```diff
@@ -37,45 +37,43 @@ class LocallyOfFiniteType (f : X ⟶ Y) : Prop where
...
+  obtain ⟨ι, _, f, hf⟩ := Algebra.FiniteType.iff_quotient_mvPolynomial'.mp H
+  let g := (Algebra.TensorProduct.productLeftAlgHom
+    (Algebra.TensorProduct.includeLeft : S →ₐ[S] S ⊗[R] T)
+    (Algebra.TensorProduct.includeRight.comp f))
+  have hg : Function.Surjective g := by
+    show Function.Surjective (g.toLinearMap.restrictScalars R)
+    convert f.toLinearMap.lTensor_surjective S hf using 2
+    ext; simp [g]
+  exact ((Algebra.FiniteType.mvPolynomial _ _).equiv
+    (MvPolynomial.algebraTensorAlgEquiv R (σ := ι) S).symm).of_surjective _ hg
```
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Morphisms/FiniteType.lean:77]:
```suggestion
lemma locallyOfFiniteType_stableUnderBaseChange :
    MorphismProperty.StableUnderBaseChange @LocallyOfFiniteType :=
  HasRingHomProperty.stableUnderBaseChange RingHom.finiteType_stableUnderBaseChange
```

File: Mathlib/AlgebraicGeometry/Morphisms/Constructors.lean [hunk 7]
```diff
@@ -60,6 +60,7 @@ instance AffineTargetMorphismProperty.diagonal_respectsIso (P : AffineTargetMorp
     rw [pullback.mapDesc_comp, P.cancel_right_of_respectsIso]
     apply H
 
+attribute [-ext] Scheme.Hom.ext in
```
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Morphisms/Constructors.lean:63]: Is this an indication that the `@[ext]` on `Scheme.Hom.ext` should be dropped?

File: Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.lean [hunk 8]
```diff
@@ -38,504 +61,311 @@ universe u
...
+  exact f.appLE_congr _ (ιOpens_image_top _) congr($(this).1.1) _
+
+theorem affineLocally_iff_affineOpens_le {X Y : Scheme.{u}} (f : X ⟶ Y) :
+    affineLocally.{u} P f ↔
+      ∀ (U : Y.affineOpens) (V : X.affineOpens) (e : V.1 ≤ f ⁻¹ᵁ U.1), P (f.appLE U V e) :=
+  forall_congr' fun U ↦ sourceAffineLocally_morphismRestrict P f U U.2
+
+theorem sourceAffineLocally_isLocal (h₁ : RingHom.RespectsIso P)
+    (h₂ : RingHom.LocalizationPreserves P) (h₃ : RingHom.OfLocalizationSpan P) :
+    (sourceAffineLocally P).IsLocal := by
```
[the remark this answers is not in the document]
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.lean:142]: `RingHom.PropertyIsLocal` contains precisely what is needed for `affineLocally` to be local at the source and the target. `AffineTargetMorphismProperty.IsLocal` only guarantees that it is local at the target.

File: Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.lean [hunk 9]
```diff
@@ -38,504 +61,311 @@ universe u
...
+
+end affineLocally
+
+/--
+`HasRingHomProperty P Q` is a type class asserting that `P` is local at the target and the source,
+and for `f : Spec B ⟶ Spec A`, it is equivalent to the ring hom property `Q`.
+To make the proofs easier, we state it instead as
+1. `Q` is local (See `RingHom.PropertyIsLocal`)
+2. `P f` if and only if `Q` holds for every `Γ(Y, U) ⟶ Γ(X, V)` for all affine `U`, `V`.
+-/
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.lean:174]:
```suggestion
See `HasRingHomProperty.iff_appLE`.
-/
```

File: Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.lean (hunk 4 above)
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.lean:321]: This code looks better but ideally we should have some automation to do it for us...

File: Mathlib/AlgebraicGeometry/Morphisms/Constructors.lean (hunk 7 above)
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Morphisms/Constructors.lean:63]: I think when the user invokes `ext` when the goal is the equality of two morphisms, `Scheme.Hom.ext` is probably what one would expect, and we should allow this behavior. In this case I could instead change the `ext` to an `ext1` (which is what I should do instead) so I don't think this `ext`-lemma is that intrusive.

File: Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.lean [hunk 10]
```diff
@@ -38,504 +61,311 @@ universe u
...
+class HasRingHomProperty (P : MorphismProperty Scheme.{u})
+    (Q : outParam (∀ {R S : Type u} [CommRing R] [CommRing S], (R →+* S) → Prop)) : Prop where
```
[the remark this answers is not in the document]
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.lean:177]: Sounds like a good idea. Will do it in a separate PR.

File: Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.lean [hunk 11]
```diff
@@ -27,8 +27,31 @@ the target. (TODO)
...
+- `AlgebraicGeometry.HasAffineProperty.iff_of_isAffine`:
+    If `X` and `Y` are affine, then `P f ↔ Q (f.app ⊤)`.
+- `AlgebraicGeometry.HasAffineProperty.Spec_iff`:
+    `P (Spec.map φ) ↔ Q φ`
+- `AlgebraicGeometry.HasAffineProperty.iff_of_iSup_eq_top`:
+    If `Y` is affine, `P f ↔ ∀ i, Q (f.appLE ⊤ (U i) _)` for a family `U` of affine opens of `X`.
+- `AlgebraicGeometry.HasAffineProperty.of_isOpenImmersion`:
+    If `f` is an open immersion then `P f`.
+- `AlgebraicGeometry.HasAffineProperty.stableUnderBaseChange`:
+    If `Q` is stable under base change, then so is `P`.
```
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.lean:51]: These still have wrong names, `HasAffineProperty` -> `HasRingHomProperty`.

**Review by chrisflav — APPROVED** [review]
Thanks! Apart from one comment, LGTM!

**Comment by mattrobball** [comment]
!bench

File: Mathlib/AlgebraicGeometry/Restrict.lean [hunk 12]
```diff
@@ -365,6 +365,16 @@ theorem Γ_map_morphismRestrict {X Y : Scheme.{u}} (f : X ⟶ Y) (U : Opens Y) :
...
   rfl
 
+@[simp]
+lemma morphismRestrict_id {X : Scheme.{u}} (U : Opens X) : 𝟙 X ∣_ U = 𝟙 _ := by
+  ext1
+  · ext; erw [morphismRestrict_val_base (𝟙 X) U]; rfl
+  · simp only [Scheme.restrict_presheaf_obj, Scheme.id_val_base, Opens.carrier_eq_coe,
+      TopCat.coe_id, id_eq, eq_mpr_eq_cast, morphismRestrict_app', Scheme.Hom.appLE, Scheme.id_app,
+      Category.id_comp, eqToHom_op, Scheme.restrict_presheaf_map, ← Functor.map_comp]
+    rw [← X.presheaf.map_id]
```
Comment (mattrobball) [inline Mathlib/AlgebraicGeometry/Restrict.lean:375]: This line is responsible for all the regression in the file. Tracing it crashes the LSP.

**Comment by erdOne** [comment]
!bench

File: Mathlib/AlgebraicGeometry/Restrict.lean (hunk 12 above)
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Restrict.lean]: I changed it into a new proof that should be a lot faster. Let's see how the new benchmark says.

Comment (mattrobball) [reply Mathlib/AlgebraicGeometry/Restrict.lean]: Great! It certainly looks much cleaner

**Comment by erdOne** [comment]
Okay the deletion of those comments seems to have pissed off CI... not sure how to fix it.

**Comment by mattrobball** [comment]
Can you add a PR description saying a bit about the generalization?

**Comment by erdOne** [comment]
Done.

**Comment by mattrobball** [comment]
bors merge

**Pull request closed without merging**